File: ai_features.py
"""
AI Features Module for Anvesh
Includes OCR, Object Detection, Face Detection, and Face Matching
"""
import os
import cv2
import numpy as np
from typing import List, Dict, Optional, Tuple
import base64
from PIL import Image
import io
import json
from pathlib import Path

# Try to import AI libraries
try:
    import pytesseract
    TESSERACT_AVAILABLE = True
except ImportError:
    TESSERACT_AVAILABLE = False

try:
    import face_recognition
    FACE_RECOGNITION_AVAILABLE = True
except ImportError:
    FACE_RECOGNITION_AVAILABLE = False

try:
    from ultralytics import YOLO
    YOLO_AVAILABLE = True
except ImportError:
    YOLO_AVAILABLE = False

class AIFeatures:
    """AI-powered features for Anvesh"""
    
    def __init__(self):
        self.face_encodings_cache = {}
        self.yolo_model = None
        self._load_models()
    
    def _load_models(self):
        """Load AI models"""
        if YOLO_AVAILABLE:
            try:
                # Use YOLOv8n (nano) for speed, can be upgraded to YOLOv8s/m/l/x
                self.yolo_model = YOLO('yolov8n.pt')  # Will download on first use
            except Exception as e:
                print(f"Warning: Could not load YOLO model: {e}")
                self.yolo_model = None
    
    def extract_text_from_image(self, image_path: str) -> Dict:
        """Extract text from image using OCR"""
        if not TESSERACT_AVAILABLE:
            return {"error": "Tesseract OCR not available. Install: pip install pytesseract"}
        
        try:
            image = cv2.imread(image_path)
            if image is None:
                return {"error": "Could not read image"}
            
            # Preprocess image for better OCR
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
            # Apply thresholding
            _, thresh = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
            
            # Extract text
            text = pytesseract.image_to_string(thresh, lang='eng')
            data = pytesseract.image_to_data(thresh, output_type=pytesseract.Output.DICT)
            
            # Get bounding boxes
            boxes = []
            n_boxes = len(data['text'])
            for i in range(n_boxes):
                if int(data['conf'][i]) > 0:
                    boxes.append({
                        'text': data['text'][i],
                        'confidence': float(data['conf'][i]),
                        'left': data['left'][i],
                        'top': data['top'][i],
                        'width': data['width'][i],
                        'height': data['height'][i]
                    })
            
            return {
                "success": True,
                "text": text.strip(),
                "boxes": boxes,
                "full_text": text
            }
        except Exception as e:
            return {"error": str(e)}
    
    def extract_text_from_video(self, video_path: str, frame_interval: int = 30) -> Dict:
        """Extract text from video frames using OCR"""
        if not TESSERACT_AVAILABLE:
            return {"error": "Tesseract OCR not available"}
        
        try:
            cap = cv2.VideoCapture(video_path)
            if not cap.isOpened():
                return {"error": "Could not open video"}
            
            frame_count = 0
            all_text = []
            frame_texts = []
            
            while True:
                ret, frame = cap.read()
                if not ret:
                    break
                
                # Process every Nth frame
                if frame_count % frame_interval == 0:
                    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
                    _, thresh = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
                    
                    text = pytesseract.image_to_string(thresh, lang='eng')
                    if text.strip():
                        all_text.append(text.strip())
                        frame_texts.append({
                            "frame": frame_count,
                            "time": frame_count / cap.get(cv2.CAP_PROP_FPS),
                            "text": text.strip()
                        })
                
                frame_count += 1
            
            cap.release()
            
            return {
                "success": True,
                "total_frames": frame_count,
                "frames_processed": len(frame_texts),
                "text": "\n\n".join(all_text),
                "frame_texts": frame_texts
            }
        except Exception as e:
            return {"error": str(e)}
    
    def detect_objects(self, image_path: str, confidence_threshold: float = 0.25) -> Dict:
        """Detect objects in image using YOLO"""
        if not YOLO_AVAILABLE or self.yolo_model is None:
            return {"error": "YOLO model not available"}
        
        try:
            results = self.yolo_model(image_path, conf=confidence_threshold)
            
            detections = []
            for result in results:
                boxes = result.boxes
                for box in boxes:
                    cls = int(box.cls[0])
                    conf = float(box.conf[0])
                    xyxy = box.xyxy[0].cpu().numpy()
                    
                    detections.append({
                        "class": result.names[cls],
                        "confidence": conf,
                        "bbox": {
                            "x1": float(xyxy[0]),
                            "y1": float(xyxy[1]),
                            "x2": float(xyxy[2]),
                            "y2": float(xyxy[3])
                        }
                    })
            
            return {
                "success": True,
                "detections": detections,
                "count": len(detections)
            }
        except Exception as e:
            return {"error": str(e)}
    
    def detect_faces(self, image_path: str) -> Dict:
        """Detect faces in image"""
        if not FACE_RECOGNITION_AVAILABLE:
            return {"error": "Face recognition not available"}
        
        try:
            image = face_recognition.load_image_file(image_path)
            face_locations = face_recognition.face_locations(image)
            face_encodings = face_recognition.face_encodings(image, face_locations)
            
            faces = []
            for i, (top, right, bottom, left) in enumerate(face_locations):
                faces.append({
                    "face_id": i,
                    "location": {
                        "top": int(top),
                        "right": int(right),
                        "bottom": int(bottom),
                        "left": int(left)
                    },
                    "encoding": face_encodings[i].tolist() if i < len(face_encodings) else None
                })
            
            return {
                "success": True,
                "faces": faces,
                "count": len(faces)
            }
        except Exception as e:
            return {"error": str(e)}
    
    def detect_faces_in_video(self, video_path: str, frame_interval: int = 30) -> Dict:
        """Detect faces in video"""
        if not FACE_RECOGNITION_AVAILABLE:
            return {"error": "Face recognition not available"}
        
        try:
            cap = cv2.VideoCapture(video_path)
            if not cap.isOpened():
                return {"error": "Could not open video"}
            
            fps = cap.get(cv2.CAP_PROP_FPS)
            frame_count = 0
            all_faces = []
            face_tracking = {}  # Track faces across frames
            
            while True:
                ret, frame = cap.read()
                if not ret:
                    break
                
                if frame_count % frame_interval == 0:
                    # Convert BGR to RGB
                    rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
                    
                    face_locations = face_recognition.face_locations(rgb_frame)
                    face_encodings = face_recognition.face_encodings(rgb_frame, face_locations)
                    
                    for encoding, location in zip(face_encodings, face_locations):
                        all_faces.append({
                            "frame": frame_count,
                            "time": frame_count / fps,
                            "location": {
                                "top": int(location[0]),
                                "right": int(location[1]),
                                "bottom": int(location[2]),
                                "left": int(location[3])
                            },
                            "encoding": encoding.tolist()
                        })
                
                frame_count += 1
            
            cap.release()
            
            return {
                "success": True,
                "total_frames": frame_count,
                "faces_detected": len(all_faces),
                "faces": all_faces
            }
        except Exception as e:
            return {"error": str(e)}
    
    def match_faces(self, image1_path: str, image2_path: str, threshold: float = 0.6) -> Dict:
        """Match faces between two images"""
        if not FACE_RECOGNITION_AVAILABLE:
            return {"error": "Face recognition not available"}
        
        try:
            # Load and encode faces from both images
            img1 = face_recognition.load_image_file(image1_path)
            img2 = face_recognition.load_image_file(image2_path)
            
            encodings1 = face_recognition.face_encodings(img1)
            encodings2 = face_recognition.face_encodings(img2)
            
            if len(encodings1) == 0:
                return {"error": "No faces found in first image"}
            if len(encodings2) == 0:
                return {"error": "No faces found in second image"}
            
            # Compare all faces
            matches = []
            for i, enc1 in enumerate(encodings1):
                for j, enc2 in enumerate(encodings2):
                    distance = face_recognition.face_distance([enc1], enc2)[0]
                    similarity = (1 - distance) * 100  # Convert to percentage
                    
                    matches.append({
                        "face1_index": i,
                        "face2_index": j,
                        "distance": float(distance),
                        "similarity_percentage": float(similarity),
                        "is_match": distance < threshold
                    })
            
            return {
                "success": True,
                "faces_in_image1": len(encodings1),
                "faces_in_image2": len(encodings2),
                "matches": matches,
                "best_match": max(matches, key=lambda x: x["similarity_percentage"]) if matches else None
            }
        except Exception as e:
            return {"error": str(e)}
    
    def find_matching_faces_in_folder(self, reference_image_path: str, folder_path: str, threshold: float = 0.6) -> Dict:
        """Find all images in folder that contain matching faces"""
        if not FACE_RECOGNITION_AVAILABLE:
            return {"error": "Face recognition not available"}
        
        try:
            # Load reference face
            ref_img = face_recognition.load_image_file(reference_image_path)
            ref_encodings = face_recognition.face_encodings(ref_img)
            
            if len(ref_encodings) == 0:
                return {"error": "No faces found in reference image"}
            
            ref_encoding = ref_encodings[0]  # Use first face
            
            # Supported image formats
            image_extensions = {'.jpg', '.jpeg', '.png', '.bmp', '.gif', '.webp'}
            matches = []
            
            for root, dirs, files in os.walk(folder_path):
                for file in files:
                    if Path(file).suffix.lower() in image_extensions:
                        image_path = os.path.join(root, file)
                        try:
                            img = face_recognition.load_image_file(image_path)
                            encodings = face_recognition.face_encodings(img)
                            
                            for encoding in encodings:
                                distance = face_recognition.face_distance([ref_encoding], encoding)[0]
                                similarity = (1 - distance) * 100
                                
                                if distance < threshold:
                                    matches.append({
                                        "image_path": image_path,
                                        "similarity_percentage": float(similarity),
                                        "distance": float(distance)
                                    })
                                    break  # Found a match, move to next image
                        except Exception as e:
                            continue
            
            # Sort by similarity
            matches.sort(key=lambda x: x["similarity_percentage"], reverse=True)
            
            return {
                "success": True,
                "reference_image": reference_image_path,
                "matches_found": len(matches),
                "matches": matches
            }
        except Exception as e:
            return {"error": str(e)}
    
    def get_capabilities(self) -> Dict:
        """Get available AI capabilities"""
        return {
            "ocr": TESSERACT_AVAILABLE,
            "face_detection": FACE_RECOGNITION_AVAILABLE,
            "object_detection": YOLO_AVAILABLE and self.yolo_model is not None,
            "face_matching": FACE_RECOGNITION_AVAILABLE
        }

# Global instance
ai_features = AIFeatures()

File: static/ai_features.js
const API_BASE = 'http://127.0.0.1:8000';

document.addEventListener('DOMContentLoaded', function() {
    loadCapabilities();
});

function loadCapabilities() {
    fetch(`${API_BASE}/api/ai/capabilities`)
        .then(response => response.json())
        .then(data => {
            const statusDiv = document.getElementById('capabilitiesStatus');
            if (data.available === false) {
                statusDiv.innerHTML = `
                    <div class="alert alert-warning">
                        <i class="fas fa-exclamation-triangle me-2"></i>
                        AI features are not available. Please install required dependencies.
                        <br><small>Run: pip install opencv-python pytesseract face-recognition ultralytics</small>
                    </div>
                `;
                return;
            }
            
            const capabilities = [
                { name: 'OCR', available: data.ocr, icon: 'fa-image' },
                { name: 'Face Detection', available: data.face_detection, icon: 'fa-user' },
                { name: 'Object Detection', available: data.object_detection, icon: 'fa-eye' },
                { name: 'Face Matching', available: data.face_matching, icon: 'fa-user-friends' }
            ];
            
            const badges = capabilities.map(cap => `
                <span class="badge ${cap.available ? 'bg-success' : 'bg-secondary'} me-2 mb-2">
                    <i class="fas ${cap.icon} me-1"></i>${cap.name}: ${cap.available ? 'Available' : 'Not Available'}
                </span>
            `).join('');
            
            statusDiv.innerHTML = badges;
        })
        .catch(error => {
            document.getElementById('capabilitiesStatus').innerHTML = 
                `<div class="alert alert-danger">Error loading capabilities: ${error.message}</div>`;
        });
}

function showProcessing(message = 'Processing...') {
    document.getElementById('processingMessage').textContent = message;
    document.getElementById('processingOverlay').style.display = 'flex';
}

function hideProcessing() {
    document.getElementById('processingOverlay').style.display = 'none';
}

function showResults(content) {
    document.getElementById('resultsContainer').innerHTML = content;
    document.getElementById('resultsSection').classList.remove('d-none');
    document.getElementById('resultsSection').scrollIntoView({ behavior: 'smooth' });
}

function clearResults() {
    document.getElementById('resultsContainer').innerHTML = '';
    document.getElementById('resultsSection').classList.add('d-none');
}

function openOCRImage() {
    const path = prompt('Enter image file path:');
    if (!path) return;
    
    showProcessing('Extracting text from image...');
    
    fetch(`${API_BASE}/api/ai/ocr/image`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ file_path: path })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Extracted Text:</h6>`;
        html += `<div class="alert alert-info"><pre style="white-space: pre-wrap;">${escapeHtml(data.text || data.full_text || 'No text found')}</pre></div>`;
        
        if (data.boxes && data.boxes.length > 0) {
            html += `<h6>Text Boxes (${data.boxes.length}):</h6>`;
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Text</th><th>Confidence</th><th>Position</th></tr></thead><tbody>`;
            data.boxes.forEach(box => {
                html += `<tr><td>${escapeHtml(box.text)}</td><td>${box.confidence.toFixed(1)}%</td><td>(${box.left}, ${box.top})</td></tr>`;
            });
            html += `</tbody></table></div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openOCRVideo() {
    const path = prompt('Enter video file path:');
    if (!path) return;
    
    const interval = prompt('Process every Nth frame (default: 30):', '30');
    const frameInterval = parseInt(interval) || 30;
    
    showProcessing('Extracting text from video frames... This may take a while.');
    
    fetch(`${API_BASE}/api/ai/ocr/video`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ file_path: path, frame_interval: frameInterval })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Video OCR Results:</h6>`;
        html += `<p>Total Frames: ${data.total_frames}, Processed: ${data.frames_processed}</p>`;
        html += `<div class="alert alert-info"><pre style="white-space: pre-wrap;">${escapeHtml(data.text || 'No text found')}</pre></div>`;
        
        if (data.frame_texts && data.frame_texts.length > 0) {
            html += `<h6>Text by Frame:</h6>`;
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Frame</th><th>Time (s)</th><th>Text</th></tr></thead><tbody>`;
            data.frame_texts.forEach(item => {
                html += `<tr><td>${item.frame}</td><td>${item.time.toFixed(2)}</td><td>${escapeHtml(item.text)}</td></tr>`;
            });
            html += `</tbody></table></div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openObjectDetection() {
    const path = prompt('Enter image file path:');
    if (!path) return;
    
    const confidence = prompt('Confidence threshold (0.0-1.0, default: 0.25):', '0.25');
    const conf = parseFloat(confidence) || 0.25;
    
    showProcessing('Detecting objects in image...');
    
    fetch(`${API_BASE}/api/ai/detect/objects`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ file_path: path, confidence: conf })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Object Detection Results:</h6>`;
        html += `<p><strong>Objects Found: ${data.count}</strong></p>`;
        
        if (data.detections && data.detections.length > 0) {
            // Group by class
            const grouped = {};
            data.detections.forEach(det => {
                if (!grouped[det.class]) grouped[det.class] = [];
                grouped[det.class].push(det);
            });
            
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Object</th><th>Count</th><th>Avg Confidence</th></tr></thead><tbody>`;
            
            Object.keys(grouped).forEach(cls => {
                const items = grouped[cls];
                const avgConf = items.reduce((sum, d) => sum + d.confidence, 0) / items.length;
                html += `<tr><td><strong>${escapeHtml(cls)}</strong></td><td>${items.length}</td><td>${(avgConf * 100).toFixed(1)}%</td></tr>`;
            });
            
            html += `</tbody></table></div>`;
            
            html += `<h6>Detailed Detections:</h6>`;
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Object</th><th>Confidence</th><th>Bounding Box</th></tr></thead><tbody>`;
            data.detections.forEach(det => {
                html += `<tr><td>${escapeHtml(det.class)}</td><td>${(det.confidence * 100).toFixed(1)}%</td><td>(${det.bbox.x1.toFixed(0)}, ${det.bbox.y1.toFixed(0)}) - (${det.bbox.x2.toFixed(0)}, ${det.bbox.y2.toFixed(0)})</td></tr>`;
            });
            html += `</tbody></table></div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openFaceDetection() {
    const path = prompt('Enter image file path:');
    if (!path) return;
    
    showProcessing('Detecting faces in image...');
    
    fetch(`${API_BASE}/api/ai/detect/faces`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ file_path: path })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Face Detection Results:</h6>`;
        html += `<p><strong>Faces Found: ${data.count}</strong></p>`;
        
        if (data.faces && data.faces.length > 0) {
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Face ID</th><th>Location</th></tr></thead><tbody>`;
            data.faces.forEach(face => {
                html += `<tr><td>${face.face_id}</td><td>Top: ${face.location.top}, Right: ${face.location.right}, Bottom: ${face.location.bottom}, Left: ${face.location.left}</td></tr>`;
            });
            html += `</tbody></table></div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openFaceDetectionVideo() {
    const path = prompt('Enter video file path:');
    if (!path) return;
    
    const interval = prompt('Process every Nth frame (default: 30):', '30');
    const frameInterval = parseInt(interval) || 30;
    
    showProcessing('Detecting faces in video... This may take a while.');
    
    fetch(`${API_BASE}/api/ai/detect/faces/video`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ file_path: path, frame_interval: frameInterval })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Face Detection Results:</h6>`;
        html += `<p>Total Frames: ${data.total_frames}, Faces Detected: ${data.faces_detected}</p>`;
        
        if (data.faces && data.faces.length > 0) {
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Frame</th><th>Time (s)</th><th>Location</th></tr></thead><tbody>`;
            data.faces.forEach(face => {
                html += `<tr><td>${face.frame}</td><td>${face.time.toFixed(2)}</td><td>Top: ${face.location.top}, Right: ${face.location.right}, Bottom: ${face.location.bottom}, Left: ${face.location.left}</td></tr>`;
            });
            html += `</tbody></table></div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openFaceMatching() {
    const path1 = prompt('Enter first image file path:');
    if (!path1) return;
    
    const path2 = prompt('Enter second image file path:');
    if (!path2) return;
    
    const threshold = prompt('Similarity threshold (0.0-1.0, default: 0.6):', '0.6');
    const thresh = parseFloat(threshold) || 0.6;
    
    showProcessing('Matching faces between images...');
    
    fetch(`${API_BASE}/api/ai/match/faces`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ image1_path: path1, image2_path: path2, threshold: thresh })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Face Matching Results:</h6>`;
        html += `<p>Faces in Image 1: ${data.faces_in_image1}, Faces in Image 2: ${data.faces_in_image2}</p>`;
        
        if (data.matches && data.matches.length > 0) {
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Face 1</th><th>Face 2</th><th>Similarity</th><th>Match</th></tr></thead><tbody>`;
            data.matches.forEach(match => {
                const badgeClass = match.is_match ? 'bg-success' : 'bg-secondary';
                html += `<tr>
                    <td>${match.face1_index}</td>
                    <td>${match.face2_index}</td>
                    <td><span class="badge ${badgeClass} face-match-badge">${match.similarity_percentage.toFixed(2)}%</span></td>
                    <td>${match.is_match ? '<span class="badge bg-success">Match</span>' : '<span class="badge bg-secondary">No Match</span>'}</td>
                </tr>`;
            });
            html += `</tbody></table></div>`;
            
            if (data.best_match) {
                html += `<div class="alert alert-info mt-3">`;
                html += `<strong>Best Match:</strong> Face ${data.best_match.face1_index} ↔ Face ${data.best_match.face2_index} `;
                html += `<span class="badge bg-primary">${data.best_match.similarity_percentage.toFixed(2)}% similar</span>`;
                html += `</div>`;
            }
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function openFindFacesInFolder() {
    const refPath = prompt('Enter reference image file path:');
    if (!refPath) return;
    
    const folderPath = prompt('Enter folder path to search:');
    if (!folderPath) return;
    
    const threshold = prompt('Similarity threshold (0.0-1.0, default: 0.6):', '0.6');
    const thresh = parseFloat(threshold) || 0.6;
    
    showProcessing('Searching for matching faces in folder... This may take a while.');
    
    fetch(`${API_BASE}/api/ai/match/faces/folder`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ reference_image_path: refPath, folder_path: folderPath, threshold: thresh })
    })
    .then(response => response.json())
    .then(data => {
        hideProcessing();
        if (data.error) {
            showResults(`<div class="alert alert-danger">${data.error}</div>`);
            return;
        }
        
        let html = `<h6>Face Search Results:</h6>`;
        html += `<p><strong>Reference Image:</strong> ${escapeHtml(data.reference_image)}</p>`;
        html += `<p><strong>Matches Found: ${data.matches_found}</strong></p>`;
        
        if (data.matches && data.matches.length > 0) {
            html += `<div class="table-responsive"><table class="table table-sm">`;
            html += `<thead><tr><th>Image</th><th>Similarity</th></tr></thead><tbody>`;
            data.matches.forEach(match => {
                html += `<tr>
                    <td>${escapeHtml(match.image_path)}</td>
                    <td><span class="badge bg-success face-match-badge">${match.similarity_percentage.toFixed(2)}%</span></td>
                </tr>`;
            });
            html += `</tbody></table></div>`;
        } else {
            html += `<div class="alert alert-warning">No matching faces found.</div>`;
        }
        
        showResults(html);
    })
    .catch(error => {
        hideProcessing();
        showResults(`<div class="alert alert-danger">Error: ${error.message}</div>`);
    });
}

function escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
}

